README
AEOLUS 0.9.1   RELEASE NOTES   27/06/2017
============================
Bugfix release.

* Fixed segfault during startup. Known to
  upstream but no new version was released. 
  See https://bugs.archlinux.org/task/52761

AEOLUS 0.9.0   RELEASE NOTES   30/06/2013
============================

Maintenance release.

* Replaced libalsadrv by libzita-alsa-pcmi
* Changed license to GPL3


AEOLUS 0.8.4   RELEASE NOTES   14/03/2010
============================

* Five new temperaments provided by Hanno Hoffstadt
  and Adam Sampson.
* Added -s option to select Jack server.
* Now using libclalsdrv-2.0.0.
* Compiles without warnings with gcc-4.4.3.


AEOLUS 0.8.0   RELEASE NOTES   29/01/2008
============================

**** Experimental release. ****

* Requires JACK >= 0.109
* Receives MIDI over JACK and via ALSA sequencer.
* -C option has been removed, use AmbDec instead.
* -B option via ALSA removed.



AEOLUS 0.6.6   RELEASE NOTES
============================


1. The stops directory
----------------------

The aeolus binary itself is a generic organ synth and
does not define any instrument. In order to use Aeolus
you need a directory containing definitions of stops
and of an instrument. The one supplied with the current
release has this structure:


 stops-0.3.0 
     |
     |___ Aeolus
     |        |          
     |        |____ definition
     |        |____ presets
     |
     |___ Aeolus1
     |        |          
     |        |____ definition
     |        |____ presets
     |
     |___ Aeolus2
     |        |          
     |        |____ definition
     |        |____ presets
     |
     |___ ***.ae0
     |___ ***.ae0 
     |___ ***.ae0
     |___ ...
     |
     |___ waves
              |
              |___ ***.ae1
              |___ ***.ae1
              |___ ***.ae1
              |___ ...


The 'Aeolus' directory is the default instrument
directory. It contains two files: a 'definition'
file that specifies the layout of the organ, and
'presets' that contains registration and midi presets.

There can be more than one instrument definition
directory (in fact there are two more examples
in the supplied stops-0.3.0). These can be selected
with a command line option (see below).

The data in 'presets' only makes sense for one
particular instrument, and that's why these files
are kept together with the corresponding instrument
definition. For binary distributions, there is a
configuration option that will make Aeolus save the
presets in the users's home (see below). This will
allow the user to save presets for one instrument.

The *.ae0 files contain parameters for the additive
synthesis. There is one such file for each rank of
pipes in the organ. These are binary files and they
should not be edited. 'Power users' can use the built-in
synthesis editor to modify them (see section 5).
  
The waves directory will be empty initially. When 
Aeolus starts up, it will compute wavetables, one for
each pipe. This is indicated by the flashing stop buttons.
The same will happen whenever the tuning or temperament is
changed. These wavetables can be saved (so Aeolus will be
ready for use much faster next time), but only if the stops
directory is writeable for the user. This will not be the
case for a binary installation as the stops dir will be
system-wide (e.g. /usr/share/Aeolus/stops-0.3.0).

In order to be able to save wavetables or edited stops
the stops directory must be copied to a location where
it can be modified by the user, (e.g. ~/stops-0.3.0).


2. Run-time configuration
-------------------------

Aeolus takes run-time options from three sources:

*  the file   /etc/aeolus.conf
*  the file   ~/.aeolusrc
*  command line options

Apart from empty lines and comments (lines starting with '#')
either of the files should just contain the command line options  
you want to use, on a single line (examples given below).

If the file in the home directory exists (even empty) then the
one in /etc is not used. Options given on the command line override
those given in either file.

Command line options are:

(audio interface)

  -J     Use JACK. This is also the default. The option 
         can be used to override a -A in the files.

  -A     Use ALSA. Aeolus should work with the "default"
         device, but in recent ALSA releases this has a
         very large buffer size, and this may result in 
         excessive latency. Use of a hardware device and
         the options below is recommended. 

         Sub-options for ALSA and their defaults are:

         -d <device>             (default)
         -r <sample rate>        (48000)
         -p <period size>        (1024)
         -n <number of periods>  (2) 

(output channels)

  -C  <number of channels> 

         The default is 2 (stereo). Other legal values
         are 4 for four speakers in a square and 8 for
         eight speaker in cube setup. The built-in 
         Ambisonics decoder used for the latter two 
         formats is quite primitive and will be removed
         in future versions, to be replaced with a much
         better external application.

         Speaker order for -C 4 is clockwise starting at
         front-left. For -C 8 the first four outputs are
         the 'bottom' speakers, and the next four the
         'top' ones, each of them in the same order as
         for -C 4.

  -B     This options selects direct Ambisionics first order
         B-format output, to be used for recording or with
         an external decoder.
 
(resources)

  -S  <stops directory>

         The default is ./stops.

  -I  <instrument directory>

         This is relative to the stops directory.
         The default is 'Aeolus'.
            
  -W  <waves directory> 

         This is relative to the stops directory.
         The default is 'waves'. This options mainly exists
         for use during development and should not be used.

  -u     This option is for use with binary distributions
         only. When used, the presets file will be stored
         into the user's home directory instead of within
         the system wide instrument directory.

(general)

  -t     Selects the text mode user interface. With this
         option Aeolus does not in any way depend on X11.
         In the current version the text mode UI is just
         an empty stub and should not be used.

  -h     Prints version information and a summary of all
         command line options. 


For example if you always use Aeolus with ALSA device hw:0,
using 3 periods of 512 frames and a sample rate of 44.1 kHz,
and you have copied the stops directory to your home, then
your ~/.aeolusrc could look like this:

# Aeolus default options
-A -d hw:0 -n 2 -p 512 -r 44100 -S /home/login/stops-0.3.0

where 'login' is your login name.


3. Binary packages
------------------

This release permits packagers to install a working Aeolus
without touching a users's home directory, as follows.

*  Install the aeolus binary in /usr/bin and the two
   plugins in /usr/lib.
*  Install the stops directory in /usr/share/Aeolus.
*  Install this README into /usr/share/doc/packages/Aeolus.
*  Create the file /etc/aeolus.conf containing:

# Aeolus default options
-u -A -S /usr/share/Aeolus/stops-0.3.0

This will use the default instrument 'Aeolus', and save the
presets in .aeolus-presets in the users's home directory.


4. A quick tour of the GUI 
--------------------------

4.1 The main window

TBC

4.2 The instrument window

TBC

4.3 The audio window

TBC

4.4 The midi window

TBC


5. Editing the synthesis parameters
-----------------------------------

TBC


6. Midi control of stop buttons
-------------------------------

The protocol uses one controller number. The default is #98, but you
can change this in global.h. The message is accepted only on channels
enabled for control in the midi matrix.

The value is interpreted as follows:

  v = 01mm0ggg 

      This type of messages (bit 6 set) selects a group, and either
      resets all stops in that group or sets the mode for the second
      form below.
      
      mm = mode. This can be:
    
         00  disabled, also resets all elements in the group.
         01  set off
         10  set on
         11  toggle

      ggg = group, one of the button groups as defined in the instrument
      definition. In the GUI groups start at the top, the first one (for
      division III) being group #0.

      The values of mm and ggg are stored and need not be repeated unless
      they change.

  v = 000bbbbb

      According to the current state of mode, this command switches a
      stop on or off, or toggles its state, or does nothing at all.
      
      bbbbb = button index within the group.

      Buttons are numbered left to right, top to bottom within each
      group. The first one is #0.


EOF



debian/changelog
aeolus (0.9.1-0) unstable; urgency=medium
  *  Changed thread initialisation (Closes: #814374).

 -- Ralf Mattes <rm@mh-freiburg.de>  Tue, 27 Jun 2017 11:51:02 +0200

aeolus (0.9.0-2) unstable; urgency=medium

  * Team upload.
  * Remove myself from the Uploaders field.

 -- Alessio Treglia <alessio@debian.org>  Wed, 02 Nov 2016 22:05:59 +0000

aeolus (0.9.0-1) unstable; urgency=low

  * New upstream release.
  * Refresh patches.
  * Replace libclalsadrv-dev with libzita-alsa-pcmi.
  * Bump debhelper compatibility.
  * Bump libclxclient-dev dependency.
  * Update debian/copyright.
  * Update Vcs-* tags.
  * Bump Standards.

 -- Alessio Treglia <alessio@debian.org>  Thu, 08 Aug 2013 10:00:11 +0100

aeolus (0.8.4-6) unstable; urgency=medium

  * Replace libreadline5-dev with libreadline-dev (Closes: #634446)

 -- Alessio Treglia <alessio@debian.org>  Tue, 06 Sep 2011 00:04:36 +0200

aeolus (0.8.4-5) unstable; urgency=low

  * debian/patches/01-makefile.patch:
    - Add -pthread flag.
    - Install aeolus_*.so into /usr/lib/aeolus, they are provided for
      internal use only.
  * Switch packaging to 3.0 (quilt) format.
  * Drop unneeded 02-rename_plugin_shlibs.patch:
  * Change my email address, remove DMUA field.
  * Add git-buildpackage config file.
  * Add manpage and install it.
  * Reintroduce build-dependency on libasound2-dev to hurd and kfreebsd-*.
  * Introduce -S,-D options to shorten debian/rules file.
  * Recommends jackd.
  * Update Homepage.
  * Fix debian/watch file.
  * Adjust dh_install,dh_installdirs control files, shorten debian/rules.
  * Update debian/copyright file.
  * Remove debian/docs, AUTHORS file is unnecessary since debian/copyright
    contains all the needed information.
  * Bump Standards.

 -- Alessio Treglia <alessio@debian.org>  Wed, 13 Apr 2011 18:34:14 +0200

aeolus (0.8.4-4) unstable; urgency=medium

  * Plugins have been renamed, fix the search patch (Closes: #581714)

 -- Alessio Treglia <quadrispro@ubuntu.com>  Sat, 15 May 2010 12:16:20 +0200

aeolus (0.8.4-3) unstable; urgency=low

  * debian/patches/01-makefile.patch: Drop -march=native flag, it isn't valid
    for Debian packages as the results are unpredictable, thanks to
    Bastian Blank for reporting this (Closes: #578278).

 -- Alessio Treglia <quadrispro@ubuntu.com>  Mon, 19 Apr 2010 19:12:51 +0200

aeolus (0.8.4-2) unstable; urgency=low

  * Adjust 01-makefile.patch to prevent FTBFS with binutils-gold
    (Closes: #553630).

 -- Alessio Treglia <quadrispro@ubuntu.com>  Sun, 18 Apr 2010 12:03:44 +0200

aeolus (0.8.4-1) unstable; urgency=low

  * New upstream release.
  * Disable libasound2-dev as build-dep for kfreebsd-*,hurd-i386
    architectures.
  * Switch to debhelper 7 + quilt.
  * Rename old debian/patches/makefile.patch to 01-makefile.patch:
    - Append a SONAME to aeolus_{txt,x11}.so libraries.
  * Add README.source file to document the usage of quilt.
  * Install README file as upstream's changelog.

 -- Alessio Treglia <quadrispro@ubuntu.com>  Sat, 17 Apr 2010 20:51:51 +0200

aeolus (0.8.2-3) unstable; urgency=low

  * Create /usr/bin destination path before calling 'make install'.
  * Bump Standards.

 -- Alessio Treglia <quadrispro@ubuntu.com>  Tue, 13 Apr 2010 17:52:18 +0200

aeolus (0.8.2-2) unstable; urgency=low

  * debian/control:
    - Change Maintainer to Debian Multimedia Maintainers.
    - Add myself to Uploaders list.
    - Set DM-Upload-Allowed to yes.
    - Bump-up Standards.
    - Add Vcs-* tags.
    - Add Homepage field.
    - Build-depend on libreadline5-dev.
   * Fix debian/menu as per spec.
   * Fix desktop file as per spec.
   * debian/copyright:
     - Update according to the DEP-5 spec.
   * Remove unnecessary debian/dirs file.

 -- Alessio Treglia <quadrispro@ubuntu.com>  Tue, 05 Jan 2010 19:41:22 +0100

aeolus (0.8.2-1) unstable; urgency=low

   * New upstream version
   * Add missing build-dependency on libreadline5-dev (Closes: #564849)

 -- Free Ekanayaka <freee@debian.org>  Sat, 16 Jan 2010 15:05:00 +0100

aeolus (0.8.1-1) unstable; urgency=low

  * New upstream release
  * Updated watch file
  * Updated makefile patch

 -- Free Ekanayaka <freee@debian.org>  Mon, 14 Apr 2008 15:22:24 +0100

aeolus (0.6.6+2-4) unstable; urgency=low

  * Added icon to the .desktop file (thanks to Daniel James)

 -- Free Ekanayaka <freee@debian.org>  Mon, 21 May 2007 14:36:51 +0200

aeolus (0.6.6+2-3) unstable; urgency=low

  * Relaxed build dep on clxclient to allow etch backports

 -- Free Ekanayaka <freee@debian.org>  Thu, 17 May 2007 14:12:36 +0200

aeolus (0.6.6+2-2) unstable; urgency=low

  * Build depend on bug-fixed clxclient

 -- Free Ekanayaka <freee@debian.org>  Mon, 14 May 2007 22:18:54 +0200

aeolus (0.6.6+2-1) unstable; urgency=low

  * Initial release (Closes: #420586)

 -- Free Ekanayaka <freee@debian.org>  Mon, 23 Apr 2007 00:08:25 +0200


main.cc
/*
    Copyright (C) 2003-2008 Fons Adriaensen <fons@kokkinizita.net>
    
    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/


#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <ctype.h>
#include <string.h>
#include <sys/mman.h>
#include <signal.h>
#include <clthreads.h>
#include <dlfcn.h>
#include "audio.h"
#include "imidi.h"
#include "model.h"
#include "slave.h"
#include "iface.h"


#ifdef __linux__
static char *options = "htuAJBM:N:S:I:W:d:r:p:n:";
#else
static char *options = "htuJBM:N:S:I:W:";
#endif
static char  optline [1024];
static bool  t_opt = false;
static bool  u_opt = false;
static bool  A_opt = false;
static bool  B_opt = false;
static int   r_val = 48000;
static int   p_val = 1024;
static int   n_val = 2;
static const char *N_val = "aeolus";
static const char *S_val = "stops";
static const char *I_val = "Aeolus";
static const char *W_val = "waves";
static const char *d_val = "default";
static Lfq_u32  note_queue (256);
static Lfq_u32  comm_queue (256);
static Lfq_u8   midi_queue (1024);
static Iface   *iface;


static void help (void)
{
    fprintf (stderr, "\nAeolus %s\n\n",VERSION);
    fprintf (stderr, "  (C) 2003-2008 Fons Adriaensen  <fons@kokkinizita.net>\n");
    fprintf (stderr, "Options:\n");
    fprintf (stderr, "  -h                 Display this text\n");
    fprintf (stderr, "  -t                 Text mode user interface\n");
    fprintf (stderr, "  -u                 Use presets file in user's home dir\n");
#ifdef __linux__
    fprintf (stderr, "  -N <name>          Name to use as JACK and ALSA client [aeolus]\n");   
#else
    fprintf (stderr, "  -N <name>          Name to use as JACK and CoreMIDI client [aeolus]\n");   
#endif
    fprintf (stderr, "  -S <stops>         Name of stops directory [stops]\n");   
    fprintf (stderr, "  -I <instr>         Name of instrument directory [Aeolus]\n");   
    fprintf (stderr, "  -W <waves>         Name of waves directory [waves]\n");   
    fprintf (stderr, "  -B                 Ambisonics B format output (JACK only)\n");
    fprintf (stderr, "  -J                 Use JACK (default)\n");
#ifdef __linux__
    fprintf (stderr, "  -A                 Use ALSA, with options:\n");
    fprintf (stderr, "    -d <device>        Alsa device [default]\n");
    fprintf (stderr, "    -r <rate>          Sample frequency [48000]\n");
    fprintf (stderr, "    -p <period>        Period size [1024]\n");
    fprintf (stderr, "    -n <nfrags>        Number of fragments [2]\n\n");
#endif
    exit (1);
}


static void procoptions (int ac, char *av [], const char *where)
{
    int k;
    
    optind = 1;
    opterr = 0;
    while ((k = getopt (ac, av, options)) != -1)
    {
        if (optarg && (*optarg == '-'))
        {
            fprintf (stderr, "\n%s\n", where);
	    fprintf (stderr, "  Missing argument for '-%c' option.\n", k); 
            fprintf (stderr, "  Use '-h' to see all options.\n");
            exit (1);
        }
	switch (k)
	{
        case 'h' : help (); exit (0);
 	case 't' : t_opt = true;  break;
 	case 'u' : u_opt = true;  break;
 	case 'A' : A_opt = true;  break;
	case 'J' : A_opt = false; break;
	case 'B' : B_opt = true; break;
        case 'r' : r_val = atoi (optarg); break;
        case 'p' : p_val = atoi (optarg); break;
        case 'n' : n_val = atoi (optarg); break;
        case 'N' : N_val = optarg; break; 
        case 'S' : S_val = optarg; break; 
        case 'I' : I_val = optarg; break; 
        case 'W' : W_val = optarg; break; 
        case 'd' : d_val = optarg; break; 
        case '?':
            fprintf (stderr, "\n%s\n", where);
            if (optopt != ':' && strchr (options, optopt)) fprintf (stderr, "  Missing argument for '-%c' option.\n", optopt); 
            else if (isprint (optopt)) fprintf (stderr, "  Unknown option '-%c'.\n", optopt);
            else fprintf (stderr, "  Unknown option character '0x%02x'.\n", optopt & 255);
            fprintf (stderr, "  Use '-h' to see all options.\n");
            exit (1);
        default:
            abort ();
 	}
    }
}


static int readconfig (const char *path)
{
    FILE *F;
    char  s [1024];
    char *p;
    char *av [30];
    int   ac;

    av [0] = 0;
    ac = 1;
    if ((F = fopen (path, "r")))
    {
        while (fgets (optline, 1024, F) && (ac == 1))
        {
 	    p = strtok (optline, " \t\n");
            if (*p == '#') continue;
            while (p && (ac < 30))
	    {
		av [ac++] = p;
                p = strtok (0, " \t\n");
	    }
	}
	fclose (F);
        sprintf (s, "In file '%s':", path);
        procoptions (ac, av, s);
        return 0;
    }
    return 1;
}


static void sigint_handler (int)
{
    signal (SIGINT, SIG_IGN);
    iface->stop ();
}


int main (int ac, char *av [])
{
    ITC_ctrl       itcc;
    Audio         *audio;
    Imidi         *imidi;
    Model         *model;
    Slave         *slave;
    void          *so_handle;
    iface_cr      *so_create;
    char           s [1024];
    char          *p;
    int            n;
     
    p = getenv ("HOME");
    if (p) sprintf (s, "%s/.aeolusrc", p);
    else strcpy (s, ".aeolusrc");
    if (readconfig (s)) readconfig ("/etc/aeolus.conf"); 
    procoptions (ac, av, "On command line:");

    if (mlockall (MCL_CURRENT | MCL_FUTURE)) fprintf (stderr, "Warning: memory lock failed.\n");

    if (t_opt) sprintf (s, "%s/aeolus_txt.so", LIBDIR);
    else       sprintf (s, "%s/aeolus_x11.so", LIBDIR);
    so_handle = dlopen (s, RTLD_NOW);
    if (! so_handle)
    {
        fprintf (stderr, "Error: can't open user interface plugin: %s.\n", dlerror ());
        return 1;
    }
    so_create = (iface_cr *) dlsym (so_handle, "create_iface");
    if (! so_create)
    {
        fprintf (stderr, "Error: can't create user interface plugin: %s.\n", dlerror ());
        dlclose (so_handle);
        return 1;
    }

    audio = new Audio (N_val, &note_queue, &comm_queue);
#ifdef __linux__
    if (A_opt) audio->init_alsa (d_val, r_val, p_val, n_val);
    else       audio->init_jack (B_opt, &midi_queue);
#else
    audio->init_jack (B_opt, &midi_queue);
#endif
    model = new Model (&comm_queue, &midi_queue, audio->midimap (), audio->appname (), S_val, I_val, W_val, u_opt);
    imidi = new Imidi (&note_queue, &midi_queue, audio->midimap (), audio->appname ());
    slave = new Slave ();
    iface = so_create (ac, av);

    ITC_ctrl::connect (audio, EV_EXIT,  &itcc, EV_EXIT);    
    ITC_ctrl::connect (audio, EV_QMIDI, model, EV_QMIDI);    
    ITC_ctrl::connect (audio, TO_MODEL, model, FM_AUDIO);    
    ITC_ctrl::connect (imidi, EV_EXIT,  &itcc, EV_EXIT);    
    ITC_ctrl::connect (imidi, EV_QMIDI, model, EV_QMIDI);    
    ITC_ctrl::connect (imidi, TO_MODEL, model, FM_IMIDI);    
    ITC_ctrl::connect (model, EV_EXIT,  &itcc, EV_EXIT);    
    ITC_ctrl::connect (model, TO_AUDIO, audio, FM_MODEL);    
    ITC_ctrl::connect (model, TO_SLAVE, slave, FM_MODEL);    
    ITC_ctrl::connect (model, TO_IFACE, iface, FM_MODEL);    
    ITC_ctrl::connect (slave, EV_EXIT,  &itcc, EV_EXIT);    
    ITC_ctrl::connect (slave, TO_AUDIO, audio, FM_SLAVE);    
    ITC_ctrl::connect (slave, TO_MODEL, model, FM_SLAVE);    
    ITC_ctrl::connect (iface, EV_EXIT,  &itcc, EV_EXIT);    
    ITC_ctrl::connect (iface, TO_MODEL, model, FM_IFACE);    

    audio->start ();
    if (imidi->thr_start (SCHED_FIFO, audio->relpri () - 20, 0x00010000))
    {
        fprintf (stderr, "Warning: can't run midi thread in RT mode.\n");
	imidi->thr_start (SCHED_OTHER, 0, 0x00010000);
    }
    if (model->thr_start (SCHED_FIFO, audio->relpri () - 30, 0x00010000))
    {
        fprintf (stderr, "Warning: can't run model thread in RT mode.\n");
	model->thr_start (SCHED_OTHER, 0, 0x00010000);
    }
    slave->thr_start (SCHED_OTHER, 0, 0x00010000);
    iface->thr_start (SCHED_OTHER, 0, 0x00010000);

    signal (SIGINT, sigint_handler); 
    n = 4;
    while (n)
    {    
	itcc.get_event (1 << EV_EXIT);
	{
	    if (n-- == 4)
	    {   
                imidi->terminate ();
		model->terminate ();
		slave->terminate ();
		iface->terminate ();
	    }
	}
    }

    delete audio;
    delete imidi;
    delete model;
    delete slave;
    delete iface;
    dlclose (so_handle);
 
    return 0;
}

